graph_weather/models/atmorep/training/loss.py

```diff
@@ -0,0 +1,247 @@
+                    N = field_mask.shape[-1]
+                    grid_size = int(math.sqrt(N))
+                    spatial_mask = rearrange(
+                        field_mask.float(), "b t (x y) -> b t x y", x=grid_size, y=grid_size
+                    )
+                    spatial_mask = F.interpolate(spatial_mask, size=(H, W), mode="nearest")
+                    spatial_mask = rearrange(spatial_mask, "b t h w -> 1 b t h w")
+                else:
+                    raise ValueError("Unexpected mask dimensions.")
+            else:
+                # If no mask is provided, use a mask of all ones
+                spatial_mask = torch.ones((1, B, T, H, W), device=field_target.device)
+
+            # -------------------------------------------------
+            # Expand the target if predictions have an
+            # ensemble dimension [E, B, T, H, W].
+            # -------------------------------------------------
+            if field_preds.dim() == 5:
+                # field_preds: [E, B, T, H, W]
+                field_target_expanded = repeat(
+                    field_target, "b t h w -> e b t h w", e=field_preds.shape[0]
+                )
+            elif field_preds.dim() == 4:
+                # field_preds: [B, T, H, W]
+                field_target_expanded = field_target
+            else:
+                raise ValueError("Unexpected prediction dimensions.")
+
+            # -------------------------------------------------
+            # Compute MSE and apply mask
+            # -------------------------------------------------
+            mse = self.mse_loss(field_preds, field_target_expanded)
+            masked_mse = (mse * spatial_mask).sum() / (spatial_mask.sum() + 1e-8)
+
+            # Apply any field-specific weighting
+            weight = 1.0
+            if field_weights is not None:
+                weight = field_weights.get(field, 1.0)
+
+            field_loss = weight * masked_mse
+            field_losses[field] = field_loss
+            total_loss += field_loss
+
+        # Average over the number of predicted fields
+        avg_loss = total_loss / num_fields
+
+        return avg_loss, field_losses
+
+
+class PhysicalConsistencyLoss(nn.Module):
+    """
+    Enforces simple physical constraints among fields, e.g.:
+      - Non-negative wind speed from u10, v10
+      - Penalizes large temperature gradients in t2m
+    """
+
+    def __init__(self, grad_threshold: float = 10.0):
+        """
+        Args:
+            grad_threshold (float): Threshold for penalizing temperature
+                                    gradients. Default is 10.0.
+        """
+        super().__init__()
+        self.grad_threshold = grad_threshold
+
+    def forward(self, predictions: dict):
+        """
+        Args:
+            predictions (dict): Dict of field predictions, each with shape
+                [E, B, T, H, W] or [B, T, H, W].
```

Comment:
```suggestion
                [Ensemble, Batch, Time, Height, Width] or [Batch, Time, Height, Width].
```
For these in the docstrings, I think it is helpful to write out what the dimension ordering means, just to be a bit clearer. If you could do that on the other docstrings that would be great!